src/commands/serve.rs

```diff
@@ -176,64 +225,158 @@ impl ServeCommand {
                 state.clone(),
                 Some(client.clone()),
                 &args.undoc_args,
             )
             .await
             {
-                anyhow::bail!(
-                    "Error during initial undoc API discovery: {err:#}\n{ISSUE_76_EXPLANATION}"
-                );
-            }
-
-            // only record the client after we've completed the
-            // initial undoc disco attempt
-            state.set_undoc_client(client).await;
+                Ok(()) => {
+                    api_discovery_ok = true;
+                    // only record the client after we've completed the
+                    // initial undoc disco attempt
+                    state.set_undoc_client(client).await;
 
-            // spawn periodic discovery task
-            let state = state.clone();
-            let args = args.undoc_args.clone();
-            tokio::spawn(async move {
-                loop {
-                    sleep(Duration::from_secs(600)).await;
-                    if let Err(err) =
-                        enumerate_devices_via_undo_api(state.clone(), None, &args).await
-                    {
-                        log::error!("Error during periodic undoc API discovery: {err:#}");
+                    // spawn periodic discovery task
+                    let state = state.clone();
+                    let args = args.undoc_args.clone();
+                    tokio::spawn(async move {
+                        loop {
+                            sleep(Duration::from_secs(600)).await;
+                            if let Err(err) =
+                                enumerate_devices_via_undo_api(state.clone(), None, &args).await
+                            {
+                                log::error!("Error during periodic undoc API discovery: {err:#}");
+                            }
+                        }
+                    });
+                }
+                Err(err) => {
+                    // Check if we can fall back to cached data
+                    if cached_device_count > 0 || api_discovery_ok {
+                        log::warn!("Undoc API discovery failed: {err:#}");
+                        if cached_device_count > 0 && !api_discovery_ok {
+                            log::warn!(
+                                "Continuing with {} cached devices from persistent database",
+                                cached_device_count
+                            );
+                        }
+                        // Continue with whatever data we have
+                    } else {
+                        // Fresh install with no cached data - we need API access
+                        anyhow::bail!(
+                            "Error during initial undoc API discovery: {err:#}\n\
+                            No cached device data available to fall back to.\n\
+                            {ISSUE_76_EXPLANATION}"
+                        );
```

Comment:
similar comments as the case above: I think we can move the notice to happen more centrally after all discovery is done, and remove the ISSUE_76 error case.